### PR TITLE
Lose pins and add constraints.txt

### PR DESCRIPTION
**Pull request by jdahm**
## Description

Try loosening up requirements.

### CODE REVIEW

File: setup.cfg [hunk 1]
```diff
@@ -41,17 +41,17 @@ package_dir =
...
+    numpy
     packaging>=20.0
-    pybind11~=2.5
-    tabulate~=0.8
-    typing-extensions~=3.7
+    pybind11
+    tabulate
+    typing-extensions
     # ---- eve / gtc ----
     boltons>=20.0
```
Comment (jdahm) [inline setup.cfg:56]: I assume some or maybe even most of these new dependencies can have these `>=` constraints essentially removed.

**Review by egparedes — CHANGES_REQUESTED** [review]
I'm not sure if it's a good idea to use pinned dependencies for CI. It should be enough for CI to use the existing requirement constraints in the gt4py package and the `requirements-dev.txt`, otherwise there is some requirement specification that is broken and it should be fixed. In my opinion, we should not try to avoid problems in CI by using the requirements we already know that work.

I would suggest to drop this idea and just update the version constraints with looser/more up-to-date constraints if needed (guaranteed that they work well with gt4py).

Any other opinions on this @havogt or anyone else?

**Comment by jdahm** [comment]
Admittedly I'm not an expert in diagnosing CI problems, but my hope was that we could have minimum version requirements as you say, but for the case of CI exactly pinning things, so that if problems arise we can create a virtualenv and reproduce exactly (ok not exactly, only up to the OS dependencies) the environment used. Perhaps that reproducible environment isn't really needed when the logs are good enough.

**Comment by mcgibbon** [comment]
There are a couple problems this fixes:
- gt4py's requirements are stricter than they need to be, so that there's more assurance dependencies won't change behaviors in testing. This makes it difficult to merge its requirements with other projects (like ours) whose versions conflict with those strict requirements.
- If an update to a dependency does break something, it can take a very long time (in our experience) to tell that a dependency caused this, because the version will update silently and the error will show up on a PR where other work is done

> It should be enough for CI to use the existing requirement constraints in the gt4py package and the requirements-dev.txt, otherwise there is some requirement specification that is broken and it should be fixed.

This is true. This is best tested with automatic tests that periodically run the main branch against latest versions of dependencies, to see if the dependencies have broken anything. Because this uses a passing branch, this kind of testing lets you know immediately that the dependency broke your project, and not your PR. In our project we don't do this because we have no users who might be using different package versions.

**Comment by egparedes** [comment]
> gt4py's requirements are a little stricter than they need to be, so that there's more assurance dependencies won't change behaviors in testing. This makes it difficult to merge its requirements with other projects (like ours) whose versions conflict.

I agree, and it's perfectly fine to loose up the version constraints of the requirements, but this PR basically removes all the constraints in `setup.cfg`.

> If an update to a dependency does break something, it can take a very long time (in our experience) to tell that a dependency caused this, because the version will update silently and the error will show up on a PR where other work is done

Yes, this is also true, but I don't think we have found this kind of issues very often in the past, so it's probably less of a problem for us than for you, but I definitely see your point.

To sum up my opinion regarding the PR: I think it's completely fine to make the requirements versions less strict (provided that they still work with gt4py) but strongly disagree about removing _all_ the constraints from the package definition. Adding pinned version numbers for CI could be ok if no one else has a strong opinion against it.

**Comment by mcgibbon** [comment]
> disagree about removing all the constraints from the package definition.

I agree with this, there should at least be minimum version requirements. Where we know newer versions break things, there should also be maximum requirements, but it sounds like this hasn't happened.

**Comment by jdahm** [comment]
Would it resolve the concerns if the strict pins were changed to minimum versions, e.g. `click~=7.1` to something like `click>=7.1`?

**Comment by havogt** [comment]
I wouldn't pin CI to a specific set of versions, unless, we add testing for at least minimum versions and latest versions separately (i.e. test the setup.cfg version ranges). At the current stage of the project (limited set of known users), I, personally, wouldn't bother with that now (but if someone is interested to contribute that, why not). I would just use what's picked by CI (latest, I guess?) and we will be forced to quickly do something in case of failure. To soften the problem that @mcgibbon mentions (random failures in PR), we can add a nightly run to the existing _push to master_ actions (1 line change), such that PR failures can be compared to failures in the nightly runs.

**Comment by jdahm** [comment]
Could I propose that we start by closing this PR for now and opening another that changes compatible releases `~=` with `>=` instead? I recall a time when we had a version clash when fv3 was using pytest `6`, but gt4py called for a `5.x` version. We do run a slight risk of incompatible APIs by doing this, but it would avoid such a problem in the future.

**Comment by egparedes** [comment]
> Could I propose that we start by closing this PR for now and opening another that changes compatible releases `~=` with `>=` instead? I recall a time when we had a version clash when fv3 was using pytest `6`, but gt4py called for a `5.x` version. We do run a slight risk of incompatible APIs by doing this, but it would avoid such a problem in the future.

Fine for me.
